app/src/main/java/net/aiscope/gdd_app/ui/mask/customview/MaskLayer.kt

```diff
@@ -46,171 +35,225 @@ class MaskLayer(
+            this.strokeCap = Paint.Cap.ROUND
             this.strokeWidth = strokeWidth
         }
     }
 
-    private var pathsPaintsAndStagesNames: MutableList<PathPaintAndStageName> = LinkedList()
-    private var undoPendingPaths = 0
-    private lateinit var currentPath: PointToPointPath
-    private var currentPaintDirty = false
-    var brushDiseaseStage = BrushDiseaseStage(0, "non-initialized", 0)
+    //fields lazily initialized
+    private var _width: Int? = null
+    private var _height: Int? = null
+    var currentScale = 1f
         set(value) {
-            currentPaintDirty = currentPaintDirty || field != value
             field = value
+            paintBrushPendingRecreation = true
+            paintEraserPendingRecreation = true
         }
-    private var currentStrokeWidth = PATH_STROKE
+
+    //fields depending on init of dimensions
+    private val width by lazy { _width!! }
+    private val height by lazy { _height!! }
+    private val latestChangeBitmap by lazy {
+        Bitmap.createBitmap(
+            width,
+            height,
+            Bitmap.Config.ARGB_8888
+        )
+    }
+    private val latestChangeBitmapCanvas by lazy { Canvas(latestChangeBitmap) }
+    private val currentStateBitmap by lazy {
+        Bitmap.createBitmap(
+            width,
+            height,
+            Bitmap.Config.ARGB_8888
+        )
+    }
+    private val currentStateBitmapCanvas by lazy { Canvas(currentStateBitmap) }
+
+    //fields depending on init of stage
+    private lateinit var currentDiseaseStage: BrushDiseaseStage
+    private lateinit var currentPaintBrush: Paint
+    private lateinit var currentPaintEraser: Paint
+    private lateinit var currentPaint: Paint
+
+    //init independent fields
+    private var paintBrushPendingRecreation = true
+    private var paintEraserPendingRecreation = true
+    private var pathsAndPaints: MutableList<PathAndPaint> = LinkedList()
+    private var undoPendingPaths = 0
+    private var currentPath: PointToPointPath? = null
+    private var currentMode: MaskCustomView.Mode = MaskCustomView.Mode.Draw
         set(value) {
-            currentPaintDirty = currentPaintDirty || field != value
+            require(arrayOf(MaskCustomView.Mode.Draw, MaskCustomView.Mode.Erase).contains(value))
             field = value
         }
-    private var currentPaint = Paint()
-        get() {
-            if (currentPaintDirty) cleanCurrentPaint()
-            return field
-        }
-    private lateinit var viewDimensions: Pair<Int, Int>
-    private lateinit var bitmapDimensions: Pair<Int, Int>
 
-    fun init(width: Int, height: Int) {
-        bitmapDimensions = width to height
-        scaleBrushAndTextPaints()
+    fun initDimensions(width: Int, height: Int) {
+        require(_width == null && _height == null) { "Dimensions were initialized already!" }
+        _width = width
+        _height = height
+    }
+
+    fun initDiseaseStage(diseaseStage: BrushDiseaseStage) {
```

Comment:
So, the thing is that it was actually ready to add labels with the names of the stages next to each path... I created that code because it was requested in Trello, but in the end Edu said it was not necessary... So, I was keeping that code for when we have a preferences dialog (we could add the option "show stage labels" to the preferences). Buuuuut... The erase mode could be a problem when used together with that feature (labels without path), so I removed it, but kept this this way... I've already changed it to hold just the colour, though, as a consequence of your other suggestion, so... Done! Thanks!

Reply:
Yeah, probably we'd have to start dealing with "smart objects" instead of "plain pixels" if we want to erase the text as well. Let's make that a problem for the future us :D